File: decentpaste-app/src-tauri/plugins/tauri-plugin-share-intent/ios/Sources/ShareIntentPlugin/ShareIntentPlugin.swift
import Foundation
import Tauri
import WebKit

class ShareIntentPlugin: Plugin {

    // App Group identifier - must match extension
    private static let appGroupId = "group.com.decentpaste.application"
    private static let pendingContentKey = "shareIntent_pendingContent"
    private static let pendingTimestampKey = "shareIntent_timestamp"

    // Track the timestamp of the last processed content to prevent duplicate processing
    // Using timestamp instead of content value allows same text to be shared multiple times
    private var lastProcessedTimestamp: Date?

    // MARK: - Plugin Lifecycle

    override func load(webview: WKWebView) {
        super.load(webview: webview)

        // Register for app lifecycle notifications
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(applicationDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )

        // Check for pending content from Share Extension (cold start)
        checkPendingContent()
    }

    /// Called when app becomes active (warm start, returning from background)
    @objc func applicationDidBecomeActive() {
        checkPendingContent()
    }

    // MARK: - Content Handling

    /// Check for pending content in App Group shared storage
    private func checkPendingContent() {
        guard let defaults = UserDefaults(suiteName: Self.appGroupId) else {
            print("[ShareIntentPlugin] Failed to access App Group defaults")
            return
        }

        // Check if there's pending content
        guard let content = defaults.string(forKey: Self.pendingContentKey) else {
            return
        }

        // Get the timestamp when content was stored
        guard let timestamp = defaults.object(forKey: Self.pendingTimestampKey) as? Date else {
            // No timestamp means malformed data, clear it
            clearPendingContentFromDefaults()
            return
        }

        // Expire content after 5 minutes
        let fiveMinutesAgo = Date().addingTimeInterval(-300)
        if timestamp < fiveMinutesAgo {
            clearPendingContentFromDefaults()
            return
        }

        // Use timestamp to detect if this is new content (allows same text to be shared multiple times)
        if let lastProcessed = lastProcessedTimestamp, timestamp <= lastProcessed {
            return
        }
        lastProcessedTimestamp = timestamp

        // Emit event to frontend
        trigger("share-intent-received", data: [
            "content": content,
            "source": "ios"
        ])
    }

    /// Clear pending content from shared storage
    private func clearPendingContentFromDefaults() {
        guard let defaults = UserDefaults(suiteName: Self.appGroupId) else { return }
        defaults.removeObject(forKey: Self.pendingContentKey)
        defaults.removeObject(forKey: Self.pendingTimestampKey)
        defaults.synchronize()
    }

    // MARK: - Commands

    /// Get and consume pending content
    @objc public func getPendingContent(_ invoke: Invoke) throws {
        guard let defaults = UserDefaults(suiteName: Self.appGroupId) else {
            invoke.resolve(["content": NSNull(), "source": NSNull()])
            return
        }

        if let content = defaults.string(forKey: Self.pendingContentKey) {
            // Clear after reading (consume)
            clearPendingContentFromDefaults()

            invoke.resolve([
                "content": content,
                "source": "ios"
            ])
        } else {
            invoke.resolve(["content": NSNull(), "source": NSNull()])
        }
    }

    /// Check if there's pending content
    @objc public func hasPendingContent(_ invoke: Invoke) throws {
        guard let defaults = UserDefaults(suiteName: Self.appGroupId),
              let _ = defaults.string(forKey: Self.pendingContentKey) else {
            invoke.resolve(["hasPending": false])
            return
        }
        invoke.resolve(["hasPending": true])
    }

    /// Clear pending content without processing
    @objc public func clearPendingContent(_ invoke: Invoke) throws {
        clearPendingContentFromDefaults()
        lastProcessedTimestamp = nil
        invoke.resolve([:])
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}

@_cdecl("init_plugin_share_intent")
func initPlugin() -> Plugin {
    return ShareIntentPlugin()
}

File: decentpaste-app/src-tauri/plugins/tauri-plugin-share-intent/android/src/main/kotlin/com/decentpaste/plugins/shareintent/ShareIntentPlugin.kt
package com.decentpaste.plugins.shareintent

import android.app.Activity
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.webkit.WebView
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import org.json.JSONObject

@TauriPlugin
class ShareIntentPlugin(private val activity: Activity) : Plugin(activity) {

    companion object {
        private const val TAG = "ShareIntentPlugin"
    }

    // Store pending content for retrieval by Rust/JS
    private var pendingContent: String? = null

    // Track the last processed intent to prevent duplicate processing on config changes
    // We use a hash of content + timestamp to allow same content to be shared multiple times
    private var lastProcessedIntentHash: Int = 0

    // Store webView reference for event emission
    private var webViewRef: WebView? = null

    /**
     * Called when the plugin is loaded and WebView is ready.
     * Check for share intent that launched the app (cold start).
     */
    override fun load(webView: WebView) {
        super.load(webView)
        webViewRef = webView
        Log.i(TAG, "ShareIntentPlugin loaded")
        Log.i(TAG, "Checking activity intent: action=${activity.intent?.action}, type=${activity.intent?.type}")
        // Handle intent that started the activity
        handleIntent(activity.intent)
    }

    /**
     * Called when the activity receives a new intent while running (warm start).
     * This happens when user shares to the app while it's in background.
     */
    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        Log.i(TAG, "onNewIntent received: action=${intent.action}, type=${intent.type}")
        handleIntent(intent)
    }

    /**
     * Emit an event to the frontend via JavaScript evaluation.
     * Uses window.__TAURI__.event.emit for Tauri v2 compatibility.
     */
    private fun emitEvent(eventName: String, content: String, source: String) {
        val webView = webViewRef ?: run {
            Log.e(TAG, "WebView not available for event emission")
            return
        }

        // Escape content for JavaScript string
        val escapedContent = content
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")

        // JavaScript to emit the event using Tauri's event system
        val js = """
            (function() {
                if (window.__TAURI__ && window.__TAURI__.event && window.__TAURI__.event.emit) {
                    window.__TAURI__.event.emit('$eventName', { content: "$escapedContent", source: "$source" });
                    console.log('[ShareIntentPlugin] Event emitted via __TAURI__.event.emit');
                } else if (window.__TAURI_INTERNALS__ && window.__TAURI_INTERNALS__.invoke) {
                    // Fallback for older Tauri versions
                    window.__TAURI_INTERNALS__.invoke('plugin:event|emit', { event: '$eventName', payload: { content: "$escapedContent", source: "$source" } });
                    console.log('[ShareIntentPlugin] Event emitted via __TAURI_INTERNALS__.invoke');
                } else {
                    // Last resort: dispatch a custom DOM event
                    window.dispatchEvent(new CustomEvent('$eventName', { detail: { content: "$escapedContent", source: "$source" } }));
                    console.log('[ShareIntentPlugin] Event emitted via dispatchEvent');
                }
            })();
        """.trimIndent()

        // Run on UI thread
        Handler(Looper.getMainLooper()).post {
            webView.evaluateJavascript(js) { result ->
                Log.d(TAG, "JavaScript evaluation result: $result")
            }
        }
    }

    /**
     * Process an intent and extract shared text if present.
     */
    private fun handleIntent(intent: Intent?) {
        if (intent == null) {
            Log.d(TAG, "handleIntent: intent is null")
            return
        }

        Log.d(TAG, "handleIntent: action=${intent.action}, type=${intent.type}")

        // Check if this is a share intent
        if (intent.action == Intent.ACTION_SEND && intent.type == "text/plain") {
            val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
            Log.i(TAG, "Share intent detected! Text: ${sharedText?.take(50)}...")

            if (!sharedText.isNullOrEmpty()) {
                // Use Intent's identity hash to detect if this is the same intent being reprocessed
                // (e.g., on configuration changes). This allows the same text to be shared multiple times.
                val intentHash = System.identityHashCode(intent)

                if (intentHash == lastProcessedIntentHash) {
                    Log.d(TAG, "Skipping already-processed intent (same object)")
                    return
                }
                lastProcessedIntentHash = intentHash

                // Store for retrieval via command
                pendingContent = sharedText
                Log.i(TAG, "Stored pending content (${sharedText.length} chars)")

                Log.i(TAG, "Emitting share-intent-received event via JavaScript")
                // Use a slight delay to ensure WebView is fully ready
                Handler(Looper.getMainLooper()).postDelayed({
                    emitEvent("share-intent-received", sharedText, "android")
                }, 500)
            }
        } else {
            Log.d(TAG, "Not a share intent, skipping")
        }

        // Clear the intent action to prevent re-processing on config changes
        intent.action = null
    }

    /**
     * Command to get pending share content.
     * Called from Rust/JS to retrieve and consume the pending content.
     */
    @Command
    fun getPendingContent(invoke: Invoke) {
        val result = JSObject()

        if (pendingContent != null) {
            result.put("content", pendingContent)
            result.put("source", "android")
            pendingContent = null  // Consume the content
        } else {
            result.put("content", JSONObject.NULL)
            result.put("source", JSONObject.NULL)
        }

        invoke.resolve(result)
    }

    /**
     * Command to check if there's pending content without consuming it.
     */
    @Command
    fun hasPendingContent(invoke: Invoke) {
        val result = JSObject()
        result.put("hasPending", pendingContent != null)
        invoke.resolve(result)
    }

    /**
     * Command to clear pending content without processing.
     */
    @Command
    fun clearPendingContent(invoke: Invoke) {
        pendingContent = null
        lastProcessedIntentHash = 0
        invoke.resolve(JSObject())
    }
}
